redbeams/src/main/java/com/sequenceiq/redbeams/service/network/AwsSubnetChooser.java

```diff
@@ -0,0 +1,88 @@
+package com.sequenceiq.redbeams.service.network;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Multimaps;
+import com.sequenceiq.cloudbreak.cloud.model.CloudSubnet;
+
+@Component
+public class AwsSubnetChooser {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AwsSubnetChooser.class);
+
+    private static final int MIN_SUBNET_IN_DIFFERENT_AZ = 2;
+
+    private static final int MAX_SUBNET_IN_DEFFERENT_AZ = 3;
+
+    @Inject
+    private AwsSubnetValidator awsSubnetValidator;
+
+    public List<CloudSubnet> chooseSubnets(List<CloudSubnet> subnetMetas) {
+        LOGGER.debug("Subnets to choose from: [{}]", subnetMetas);
+        awsSubnetValidator.validate(subnetMetas, MIN_SUBNET_IN_DIFFERENT_AZ);
+        return chooseAwsSubnetsInDifferentAz(subnetMetas);
+    }
+
+    private List<CloudSubnet> chooseAwsSubnetsInDifferentAz(List<CloudSubnet> subnetMetas) {
+        Map<String, Collection<CloudSubnet>> privateSubnetPerAz = collectPrivateSubnetPerAz(subnetMetas);
+        Set<CloudSubnet> privateSubnetsInDifferentAz = findMaxThreePrivateSubnetInDiffrentAz(privateSubnetPerAz);
+        List<CloudSubnet> chosenSubnets = new ArrayList<>(privateSubnetsInDifferentAz);
+        chosenSubnets.addAll(collectPublicSubnetsIfNeeded(subnetMetas, privateSubnetPerAz, chosenSubnets.size()));
+        LOGGER.debug("Chosen subnets: [{}]", chosenSubnets);
+        return chosenSubnets;
+    }
+
+    private Set<CloudSubnet> collectPublicSubnetsIfNeeded(List<CloudSubnet> subnetMetas, Map<String, Collection<CloudSubnet>> privateSubnetPerAz,
+            int chosenSubnetSize) {
+        if (chosenSubnetSize < MIN_SUBNET_IN_DIFFERENT_AZ) {
+            Map<String, Collection<CloudSubnet>> publicSubnetPerAz = mapPublicSubnetPerAz(subnetMetas, privateSubnetPerAz);
+            return collectLimitedPublicSubnetsInDifferentAz(chosenSubnetSize, publicSubnetPerAz);
+        } else {
+            LOGGER.debug("No public subnets needed");
+            return Set.of();
+        }
+    }
+
+    private Set<CloudSubnet> collectLimitedPublicSubnetsInDifferentAz(int chosenSubnetSize, Map<String, Collection<CloudSubnet>> publicSubnetPerAz) {
+        Set<CloudSubnet> chosenPublicSubnets = publicSubnetPerAz.values().stream()
+                .map(cloudSubnets -> cloudSubnets.stream().findFirst().get())
+                .limit(MIN_SUBNET_IN_DIFFERENT_AZ - chosenSubnetSize)
```

Comment:
I would limit to MAX_SUBNET_IN_DIFFERENT_AZ

Reply:
we don't want to add public subnet if we can avoid it